### PR TITLE
Fixes #28 parallel batch tracing with test

### LINKED ISSUE #28: 2.x Add ParallelBatch support
Originally submitted for v1 in #23. Was part of #24, but pulled out to be done in a separate PR.

See. 9db02da

### PR DESCRIPTION
**Pull request by llinder**
/cc @beckje01 @edwardsoo-ss

### CODE REVIEW

**Comment by edwardsoo-ss** [comment]
:+1: :cool:

File: src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java [hunk 1]
```diff
@@ -23,7 +23,12 @@ public TraceContext get() {
     return registrySupplier.get()
         .maybeGet(TraceContextHolder.class)
         .map(TraceContextHolder::getContext)
-        .orElse(null);
+        .orElseGet(() -> (Execution.isManagedThread())
+          ? Execution.current()
```
Comment (hyleung) [inline src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java]: @llinder I had a question about this from the original PR from @edwardsoo-ss - what's the difference between using `Execution.current()` and `registrySupplier.get()`? `registrySupplier` is just a method ref for `Exection::current` anyway.

Comment (llinder) [reply src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java]: This does indeed seem redundant.  The tests pass without this so I'm going to remove it.  @edwardsoo-ss is there something else we are overlooking with this?

File: src/main/java/ratpack/zipkin/TracedParallelBatch.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  static <T> TracedParallelBatch<T> of(Promise<T>... promise) {
+    return new TracedParallelBatch<>(Arrays.asList(promise));
+  }
+
+  static <T> ParallelBatch<T> of(final TraceContext context, Iterable<? extends Promise<T>> promises) {
+    return ParallelBatch.of(promises)
+        .execInit(execution ->
+            execution.add(new RatpackCurrentTraceContext.TraceContextHolder(context)));
```
Comment (llinder) [inline src/main/java/ratpack/zipkin/TracedParallelBatch.java:28]: Thinking about this more I don't think this is quite right either.

What we need to ensure is that the context is swapped out when the promise is executed and reset afterwards... similar to https://github.com/hyleung/ratpack-zipkin/blob/master/src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java#L30

Comment (llinder) [reply src/main/java/ratpack/zipkin/TracedParallelBatch.java:28]: Then again the docs seem to indicate that each batch receives a forked execution which presumably means its not ever shared.  If that is true then there doesn't seem to be a reason to worry about reseting a previous context.

File: src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java (hunk 1 above)
Comment (edwardsoo-ss) [reply src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java]: I checked and I agree that was redundant, we only needed to make the `TraceContextHolder` / `TypedValue` class to be public. my bad

File: src/main/java/ratpack/zipkin/TracedParallelBatch.java (hunk 2 above)
Comment (edwardsoo-ss) [reply src/main/java/ratpack/zipkin/TracedParallelBatch.java:28]: that is also my understanding of ParallelBatch from experimenting: 
* the forked executions do not "inherit" registry from the caller execution
* registry change in each forked execution is not shared afterwards

**Comment by hyleung** [comment]
Code change itself looks good to me (tho' can't say I've actually pulled the branch and tried it).

I'm assuming the usage won't be much different than what was described in the original PR (https://github.com/hyleung/ratpack-zipkin/pull/23#issue-219058173)?  // @llinder @edwardsoo-ss

**Comment by llinder** [comment]
@hyleung I think this is good to merge.  The test coverage should give some reassurance that it works.  As for the example usage it would look more like this now:
```
TracedParallelBatch.of(promise1, promise2)
  .withContext(parent.context())
  .yieldAll()
```

So basically works just like `ParallelBatch` except users need to pass the tracing context.  I will keep a note to update the README with an example and explanation of what happens under the covers.

**Pull request merged**